clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonBasicCredentialsLoaderSpec.groovy

```diff
@@ -18,46 +18,154 @@
-import com.netflix.spinnaker.credentials.CredentialsRepository
+import com.netflix.spinnaker.credentials.MapBackedCredentialsRepository
 import com.netflix.spinnaker.credentials.definition.CredentialsDefinitionSource
 import spock.lang.Shared
 import spock.lang.Specification
+import spock.lang.Unroll
+
+import java.util.stream.Collectors
 
 class AmazonBasicCredentialsLoaderSpec extends Specification{
-  @Shared
-  def credentialsConfig = new CredentialsConfig(){{
-    setAccessKeyId("accessKey")
-    setSecretAccessKey("secret")
-  }}
   @Shared
   def defaultAccountConfigurationProperties = new DefaultAccountConfigurationProperties()
-  @Shared
-  def credentialsRepository = Mock(CredentialsRepository) {
-    getAll() >> []
-  }
-  @Shared
-  def definitionSource = Mock(CredentialsDefinitionSource) {
-    getCredentialsDefinitions() >> []
-  }
-
-  @Shared
-  def accountsConfig = new AccountsConfiguration()
 
   def 'should set defaults'() {
+    setup:
+    def credentialsConfig = new CredentialsConfig(){{
+      setAccessKeyId("accessKey")
+      setSecretAccessKey("secret")
+    }}
+    def definitionSource = Mock(CredentialsDefinitionSource) {
+      getCredentialsDefinitions() >> []
+    }
+    def credentialsRepository = new MapBackedCredentialsRepository<NetflixAmazonCredentials>(AmazonCloudProvider.ID, null)
+    AccountsConfiguration accountsConfig = new AccountsConfiguration()
     def loader = new AmazonBasicCredentialsLoader<Account, NetflixAmazonCredentials>(
-      definitionSource, null, credentialsRepository, credentialsConfig, accountsConfig, defaultAccountConfigurationProperties
+      definitionSource,
+      null,
+      credentialsRepository,
+      credentialsConfig,
+      accountsConfig,
+      defaultAccountConfigurationProperties
     )
 
     when:
     loader.load()
 
     then:
     accountsConfig.getAccounts().size() == 1
+    with (accountsConfig.getAccounts().first()) { Account account ->
+      account.name == "default"
+      account.environment == "default"
+      account.accountType == "default"
+    }
+
     credentialsConfig.getDefaultRegions().size() == 4
     System.getProperty(SDKGlobalConfiguration.ACCESS_KEY_SYSTEM_PROPERTY) == "accessKey"
     System.getProperty(SDKGlobalConfiguration.SECRET_KEY_SYSTEM_PROPERTY) == "secret"
   }
+
+  @Unroll("should load and parse a large number of accounts with multi-threading: #multiThreadingEnabled")
+  def 'should load and parse a large number of accounts'() {
+    setup:
+    def credentialsRepository = new MapBackedCredentialsRepository<NetflixAmazonCredentials>(AmazonCloudProvider.ID, null)
+
+    // create 500 accounts
+    List<Account> accounts = new ArrayList<>()
+    for (number in 0..499) {
+      Account account = new Account(name: 'prod' + number, accountId: number)
+      if (number == 0) {
+        // test an account having a region that matches one of the default regions
```

Comment:
If this is needed to exercise the path with region sorting, please say so in the comment.

Reply:
The test is primarily meant to just parse a large number of accounts having two types of accounts - one with regions set, i.e. this one, and others without any regions set (i.e. the remaining 499 ones), which would end up using the default regions set in the credentials config or would make an AWS API call to get all the regions.

So region sorting will happen for both types of accounts. I can mention it via a comment that we need to ensure that a new list of regions is returned in the case where default regions are used - so that when multi-threading is enabled, we don't run into transient ConcurrentModificationException errors

Reply:
Yes please, since I don't see an explicit test for the ConcurrentModificationException issue, it'll be great to know what covers it.

Reply:
please see https://github.com/spinnaker/clouddriver/pull/5539/commits/4c19ceb9176959cce0ba0d37cb11fab8d30bce38

Reply:
Sorry.  There it is.